File: DFS, BFS/1707_이분 그래프.cpp
#include <iostream>
#include <vector>
#include <queue>
using namespace std;

vector <int> vertex[20001];
int visited[20001];

int bfs(int start){
    int current, next;
    queue <int> Q;
    Q.push(start);
    visited[start] = 1;

    while (Q.size() > 0){
        current = Q.front();
        Q.pop();

        for (int i = 0; i < vertex[current].size(); i++){ // 현재 정점과 연결된 모든 정점을 탐색.
            next = vertex[current][i];
            if (visited[next] == 0) { // 방문하지 않은 경우
                visited[next] = -visited[current];
                Q.push(next);
            }
            else if (visited[next] + visited[current] != 0) {
                return -1;
            }
        }
    }
    return 0;
}

int main(){
    int n, v, e, src, dst;
    int check;
    cin >> n;

    for (int i = 0; i < n; i++){
        check = 1;
        cin >> v >> e;
        for  (int j = 0; j < e; j++){
            cin >> src >> dst;
            vertex[src].push_back(dst);
            vertex[dst].push_back(src);
        }

        for (int j = 0; j < v; j++){
            visited[j+1] = 0;
        }

        for (int j = 0; j < v; j++){
            if (visited[j+1] == 0){ // 방문하지 않았을 때
                if (bfs(j+1) == -1){
                    check = -1;
                    break;
                }
            }
        }

        if (check == -1){
            cout << "NO" << endl;
        }
        else{
            cout << "YES" << endl;
        }
        
        for (int j = 0; j < v; j++){
            vertex[j+1].clear();
        }

    }
}

File: BaekJoon/DFS, BFS/1707_이분 그래프.py
import sys
from collections import deque

def bfs(vertex, visited, start):
    queue = deque()
    queue.append(start)
    visited[start] = 1

    while queue:
        current = queue.popleft()
        for element in vertex[current]:
            if visited[element] == 0:
                visited[element] = -visited[current]
                queue.append(element)
            elif visited[current] + visited[element] != 0:
                return False
    return True

def main():
    n = int(input())

    for _ in range(n):
        v, e = map(int, input().split())

        vertex = [[] for _ in range(v+1)]
        visited = [0 for _ in range(v+1)]

        check = True
        for _ in range(e):
            src, dst = map(int, sys.stdin.readline().split())
            vertex[src].append(dst)      
            vertex[dst].append(src)
        
        for i in range(v):
            if visited[i+1] == 0:
                if bfs(vertex, visited, i+1) == False:
                    check = False
                    break
        
        if check:
            print("YES")
        else:
            print("NO")

if __name__ == "__main__":
    main()
